src/draft06.rs

```diff
@@ -0,0 +1,756 @@
+use hpke::kex::KeyExchange;
+use hpke::{
+    Deserializable, EncappedKey, HpkeError, Kem as KemTrait, OpModeR, OpModeS, Serializable,
+};
+use rand::{CryptoRng, RngCore};
+use std::convert::{TryFrom, TryInto};
+use thiserror::Error as ThisError;
+
+// Extra info string used by various crypto routines.
+const LABEL_QUERY: &[u8] = b"odoh query";
+const LABEL_KEY: &[u8] = b"odoh key";
+const LABEL_NONCE: &[u8] = b"odoh nonce";
+const LABEL_KEY_ID: &[u8] = b"odoh key id";
+const LABEL_RESPONSE: &[u8] = b"odoh response";
+
+// Identifier this crate supports.
+const KEM_ID: u16 = 0x0020;
+const KDF_ID: u16 = 0x0001;
+const AEAD_ID: u16 = 0x0001;
+
+/// For the selected KDF: SHA256
+const KDF_OUTPUT_SIZE: usize = 32;
+const AEAD_KEY_SIZE: usize = 16;
+const AEAD_NONCE_SIZE: usize = 12;
+const AEAD_TAG_SIZE: usize = 16;
+
+/// This is the maximum of `AEAD_KEY_SIZE` and `AEAD_NONCE_SIZE`
+const RESPONSE_NONCE_SIZE: usize = 16;
+
+type Kem = X25519HkdfSha256;
+type Aead = AesGcm128;
+type Kdf = HkdfSha256;
+type Kex = <Kem as KemTrait>::Kex;
+
+type AeadKey = [u8; AEAD_KEY_SIZE];
+type AeadNonce = [u8; AEAD_NONCE_SIZE];
+
+/// Secret used in encrypt/decrypt API.
+pub type OdohSecret = [u8; AEAD_KEY_SIZE];
+
+/// Response nonce needed by [`encrypt_response`](fn.encrypt_response.html)
+pub type ResponseNonce = [u8; RESPONSE_NONCE_SIZE];
+
+/// HTTP content-type header required for sending queries and responses
+pub const ODOH_HTTP_HEADER: &str = "application/oblivious-dns-message";
+
+/// ODoH version supported by this library
+pub const ODOH_VERSION: u16 = 0xff06;
+
+/// Errors generated by this crate.
+#[derive(ThisError, Debug, Clone)]
+pub enum Error {
+    /// Input data is too short.
+    #[error("Input data is too short")]
+    ShortInput,
+    /// Input data has incorrect length.
+    #[error("Input data has incorrect length")]
+    InvalidInputLength,
+    /// Padding is not zero.
+    #[error("Padding is not zero")]
+    InvalidPadding,
+    /// Config parameter is invalid.
+    #[error("Config parameter is invalid")]
+    InvalidParameter,
+    /// Type byte in ObliviousDoHMessage is invalid.
+    #[error("Type byte in ObliviousDoHMessage is invalid")]
+    InvalidMessageType,
+    /// Message key_id does not match public key.
+    #[error("Message key_id does not match public key")]
+    KeyIdMismatch,
+    /// Response nonce is not equal to max(key, nonce) size.
+    #[error("Response nonce is not equal to max(key, nonce) size")]
+    InvalidResponseNonceLength,
+
+    // HpkeError doesn't support Eq
+    /// Errors from hpke crate.
+    #[error(transparent)]
+    Hpke(#[from] HpkeError),
+
+    /// Errors from aes-gcm crate.
+    #[error(transparent)]
+    AesGcm(#[from] aes_gcm::Error),
+
+    /// Unexpected internal error.
+    #[error("Unexpected internal error")]
+    Internal,
+}
+
+type Result<T, E = Error> = std::result::Result<T, E>;
+
+/// Serialize to IETF wireformat that is similar to [XDR](https://tools.ietf.org/html/rfc1014)
```

Comment:
Instead of doing all serializations by hand, you can just use https://docs.rs/bincode2/2.0.1/bincode2/fn.serialize_into.html if you want to pre allocate memory. I added more detail here: https://github.com/cloudflare/odoh-rs/pull/10#discussion_r619556666

Reply:
I'm not sure it's a good idea to use bincode for this task, reasons:
1. Using a encoding not designed for the protocol would tangle the internal data structure with binary representation, this makes it hard to change the structure later without extra attention. E.g. in the spec, `ObliviousDoHConfig` has a field `length` after `version`, but we cannot have it here as the next field is an array, which implies the length when parsed by bincode .
2. With its popularity in rust world, bincode is not so well known in other places. It's not a standard, lacks of a specification about how the serialization and deserialization  are done. Besides, we were using a even less common fork states as `A maintained fork of the now-defunct Bincode.` It feels like there's a coincidence that the library "happen" to work for this protocol.

Reply:
That's fair, no worries then.